File: DotNet/CesiumLanguageWriter/CartographicExtent.cs
using System;
using System.Diagnostics.CodeAnalysis;
using JetBrains.Annotations;

namespace CesiumLanguageWriter
{
    /// <summary>
    /// A two dimensional region specified as longitude and latitude coordinates.
    /// </summary>
    public class CartographicExtent : IEquatable<CartographicExtent>
    {
        /// <summary>
        /// Initializes a new instance with the specified extents.
        /// </summary>
        /// <param name="west">The westernmost longitude.</param>
        /// <param name="south">The southernmost latitude.</param>
        /// <param name="east">The easternmost longitude.</param>
        /// <param name="north">The northernmost latitude.</param>
        public CartographicExtent(double west, double south, double east, double north)
        {
            m_west = west;
            m_south = south;
            m_east = east;
            m_north = north;
        }

        /// <summary>
        /// Gets the northernmost latitude.
        /// </summary>
        public double NorthLatitude
        {
            get { return m_north; }
        }

        /// <summary>
        /// Gets the southernmost latitude.
        /// </summary>
        public double SouthLatitude
        {
            get { return m_south; }
        }

        /// <summary>
        /// Gets the easternmost longitude.
        /// </summary>
        public double EastLongitude
        {
            get { return m_east; }
        }

        /// <summary>
        /// Gets the westernmost longitude.
        /// </summary>
        public double WestLongitude
        {
            get { return m_west; }
        }

        /// <summary>
        /// Returns true if the specified location is inside the extent, otherwise false.
        /// </summary>
        /// <param name="longitude">The longitude.</param>
        /// <param name="latitude">The latitude.</param>
        /// <returns>
        /// <see langword="true"/> if the specified location is inside the extent (or on the border),
        /// otherwise <see langword="false"/>.
        /// </returns>
        [Pure]
        public bool IsInsideExtent(double longitude, double latitude)
        {
            return longitude >= m_west &&
                   longitude <= m_east &&
                   latitude >= m_south &&
                   latitude <= m_north;
        }

        /// <summary>
        /// Computes the union of this extent with another extent.  The union of two extents is
        /// the smallest extent that includes both.
        /// </summary>
        /// <param name="other">The other extent.</param>
        /// <returns>The union of the two extents.</returns>
        [Pure]
        public CartographicExtent Union(CartographicExtent other)
        {
            return new CartographicExtent(Math.Min(m_west, other.m_west),
                                          Math.Min(m_south, other.m_south),
                                          Math.Max(m_east, other.m_east),
                                          Math.Max(m_north, other.m_north));
        }

        /// <summary>
        /// Indicates whether another object is exactly equal to this instance.
        /// </summary>
        /// <param name="obj">The object to compare to this instance.</param>
        /// <returns><see langword="true"/> if <paramref name="obj"/> is an instance of this type and represents the same value as this instance; otherwise, <see langword="false"/>.</returns>
        public override bool Equals(object obj)
        {
            return Equals(obj as CartographicExtent);
        }

        /// <summary>
        /// Indicates whether another <see cref="CartographicExtent"/> is exactly equal to this instance.
        /// </summary>
        /// <param name="other">The <see cref="CartographicExtent"/> to compare to this instance.</param>
        /// <returns><see langword="true"/> if <paramref name="other"/> is an instance of this type and represents the same value as this instance; otherwise, <see langword="false"/>.</returns>
        [SuppressMessage("ReSharper", "CompareOfFloatsByEqualityOperator")]
        public bool Equals(CartographicExtent other)
        {
            if (ReferenceEquals(null, other))
                return false;
            if (ReferenceEquals(this, other))
                return true;

            return m_north == other.m_north &&
                   m_south == other.m_south &&
                   m_east == other.m_east &&
                   m_west == other.m_west;
        }

        /// <summary>
        /// Indicates whether each coordinate value of another instance of this type
        /// is within the required tolerance of the corresponding coordinate value of this instance.
        /// </summary>
        /// <param name="other">The set of <see cref="CartographicExtent"/> to compare to this instance.</param>
        /// <param name="epsilon">The limit at which the absolute differences between the coordinate values will not be considered equal.</param>
        /// <returns>
        /// <see langword="true"/> if the absolute differences are less than or equal to <paramref name="epsilon"/>; otherwise, <see langword="false"/>.
        /// </returns>
        [Pure]
        public bool EqualsEpsilon(CartographicExtent other, double epsilon)
        {
            return Math.Abs(m_north - other.m_north) <= epsilon &&
                   Math.Abs(m_south - other.m_south) <= epsilon &&
                   Math.Abs(m_east - other.m_east) <= epsilon &&
                   Math.Abs(m_west - other.m_west) <= epsilon;
        }

        /// <summary>
        /// Returns a hash code for this instance, which is suitable for use in hashing algorithms and data structures like a hash table.
        /// </summary>
        /// <returns>A hash code for the current object.</returns>
        public override int GetHashCode()
        {
            return HashCode.Combine(m_north.GetHashCode(),
                                    m_south.GetHashCode(),
                                    m_east.GetHashCode(),
                                    m_west.GetHashCode());
        }

        /// <summary>
        /// Returns <see langword="true"/> if the two instances are exactly equal.
        /// </summary>
        /// <param name="left">The instance to compare to <paramref name="right"/>.</param>
        /// <param name="right">The instance to compare to <paramref name="left"/>.</param>
        /// <returns>
        /// <see langword="true"/> if <paramref name="left"/> represents the same value as <paramref name="right"/>; otherwise, <see langword="false"/>.
        /// </returns>
        public static bool operator ==(CartographicExtent left, CartographicExtent right)
        {
            if (ReferenceEquals(left, null))
                return ReferenceEquals(right, null);

            return left.Equals(right);
        }

        /// <summary>
        /// Returns <see langword="true"/> if the two instances are not exactly equal.
        /// </summary>
        /// <param name="left">The instance to compare to <paramref name="right"/>.</param>
        /// <param name="right">The instance to compare to <paramref name="left"/>.</param>
        /// <returns>
        /// <see langword="true"/> if <paramref name="left"/> does not represent the same value as <paramref name="right"/>; otherwise, <see langword="false"/>.
        /// </returns>
        public static bool operator !=(CartographicExtent left, CartographicExtent right)
        {
            return !(left == right);
        }

        /// <summary>
        /// Returns the string representation of the value of this instance.
        /// </summary>
        /// <returns>
        /// A string that represents the value of this instance in the form
        /// WestLongitude, SouthLatitude, EastLongitude, NorthLatitude.
        /// </returns>
        public override string ToString()
        {
            return string.Format("{0}, {1}, {2}, {3}", m_west, m_south, m_east, m_north);
        }

        private readonly double m_north;
        private readonly double m_south;
        private readonly double m_east;
        private readonly double m_west;
    }
}

File: Java/CesiumLanguageWriter/translatedSrc/cesiumlanguagewriter/CartographicExtent.java
package cesiumlanguagewriter;


import agi.foundation.compatibility.*;
import agi.foundation.compatibility.annotations.CS2JInfo;
import agi.foundation.compatibility.annotations.CS2JWarning;
import agi.foundation.compatibility.IEquatable;
import agi.foundation.compatibility.ObjectHelper;
import agi.foundation.compatibility.PrimitiveHelper;
import agi.foundation.compatibility.StringHelper;

/**
 *  
 A two dimensional region specified as longitude and latitude coordinates.
 

 */
@SuppressWarnings( {
        "unused",
        "deprecation",
        "serial"
})
public class CartographicExtent implements IEquatable<CartographicExtent> {
    /**
    *  
    Initializes a new instance with the specified extents.
    
    
    
    
    

    * @param west The westernmost longitude.
    * @param south The southernmost latitude.
    * @param east The easternmost longitude.
    * @param north The northernmost latitude.
    */
    public CartographicExtent(double west, double south, double east, double north) {
        m_west = west;
        m_south = south;
        m_east = east;
        m_north = north;
    }

    /**
    *  Gets the northernmost latitude.
    

    */
    public final double getNorthLatitude() {
        return m_north;
    }

    /**
    *  Gets the southernmost latitude.
    

    */
    public final double getSouthLatitude() {
        return m_south;
    }

    /**
    *  Gets the easternmost longitude.
    

    */
    public final double getEastLongitude() {
        return m_east;
    }

    /**
    *  Gets the westernmost longitude.
    

    */
    public final double getWestLongitude() {
        return m_west;
    }

    /**
    *  
    Returns true if the specified location is inside the extent, otherwise false.
    
    
    
    

    * @param longitude The longitude.
    * @param latitude The latitude.
    * @return 
    {@code true} if the specified location is inside the extent (or on the border),
    otherwise {@code false}.
    
    */
    @CS2JWarning("Unhandled attribute removed: Pure")
    public final boolean isInsideExtent(double longitude, double latitude) {
        return longitude >= m_west && longitude <= m_east && latitude >= m_south && latitude <= m_north;
    }

    /**
    *  
    Computes the union of this extent with another extent.  The union of two extents is
    the smallest extent that includes both.
    
    
    

    * @param other The other extent.
    * @return The union of the two extents.
    */
    @CS2JWarning("Unhandled attribute removed: Pure")
    public final CartographicExtent union(CartographicExtent other) {
        return new CartographicExtent(Math.min(m_west, other.m_west), Math.min(m_south, other.m_south), Math.max(m_east, other.m_east), Math.max(m_north, other.m_north));
    }

    /**
    *  
    Indicates whether another object is exactly equal to this instance.
    
    
    

    * @param obj The object to compare to this instance.
    * @return {@code true} if {@code obj} is an instance of this type and represents the same value as this instance; otherwise, {@code false}.
    */
    @Override
    public boolean equals(Object obj) {
        return equalsType((obj instanceof CartographicExtent) ? (CartographicExtent) obj : null);
    }

    /**
    *  
    Indicates whether another {@link CartographicExtent} is exactly equal to this instance.
    
    
    

    * @param other The {@link CartographicExtent} to compare to this instance.
    * @return {@code true} if {@code other} is an instance of this type and represents the same value as this instance; otherwise, {@code false}.
    */
    @CS2JWarning("Unhandled attribute removed: SuppressMessage")
    public final boolean equalsType(CartographicExtent other) {
        if (ObjectHelper.referenceEquals(null, other)) {
            return false;
        }
        if (ObjectHelper.referenceEquals(this, other)) {
            return true;
        }
        return m_north == other.m_north && m_south == other.m_south && m_east == other.m_east && m_west == other.m_west;
    }

    /**
    *  
    Indicates whether each coordinate value of another instance of this type
    is within the required tolerance of the corresponding coordinate value of this instance.
    
    
    
    

    * @param other The set of {@link CartographicExtent} to compare to this instance.
    * @param epsilon The limit at which the absolute differences between the coordinate values will not be considered equal.
    * @return 
    {@code true} if the absolute differences are less than or equal to {@code epsilon}; otherwise, {@code false}.
    
    */
    @CS2JWarning("Unhandled attribute removed: Pure")
    public final boolean equalsEpsilon(CartographicExtent other, double epsilon) {
        return Math.abs(m_north - other.m_north) <= epsilon && Math.abs(m_south - other.m_south) <= epsilon && Math.abs(m_east - other.m_east) <= epsilon && Math.abs(m_west - other.m_west) <= epsilon;
    }

    /**
    *  
    Returns a hash code for this instance, which is suitable for use in hashing algorithms and data structures like a hash table.
    
    

    * @return A hash code for the current object.
    */
    @Override
    public int hashCode() {
        return HashCode.combine(PrimitiveHelper.hashCode(m_north), PrimitiveHelper.hashCode(m_south), PrimitiveHelper.hashCode(m_east), PrimitiveHelper.hashCode(m_west));
    }

    /**
    *  
    Returns {@code true} if the two instances are exactly equal.
    
    
    
    

    * @param left The instance to compare to {@code right}.
    * @param right The instance to compare to {@code left}.
    * @return 
    {@code true} if {@code left} represents the same value as {@code right}; otherwise, {@code false}.
    
    */
    @CS2JInfo("This method implements the functionality of the overloaded operator: 'System.Boolean ==(CartographicExtent,CartographicExtent)'")
    public static boolean equals(CartographicExtent left, CartographicExtent right) {
        if (ObjectHelper.referenceEquals(left, null)) {
            return ObjectHelper.referenceEquals(right, null);
        }
        return left.equalsType(right);
    }

    /**
    *  
    Returns {@code true} if the two instances are not exactly equal.
    
    
    
    

    * @param left The instance to compare to {@code right}.
    * @param right The instance to compare to {@code left}.
    * @return 
    {@code true} if {@code left} does not represent the same value as {@code right}; otherwise, {@code false}.
    
    */
    @CS2JInfo("This method implements the functionality of the overloaded operator: 'System.Boolean !=(CartographicExtent,CartographicExtent)'")
    public static boolean notEquals(CartographicExtent left, CartographicExtent right) {
        return !(CartographicExtent.equals(left, right));
    }

    /**
    *  
    Returns the string representation of the value of this instance.
    
    

    * @return 
    A string that represents the value of this instance in the form
    WestLongitude, SouthLatitude, EastLongitude, NorthLatitude.
    
    */
    @Override
    public String toString() {
        return StringHelper.format("{0}, {1}, {2}, {3}", m_west, m_south, m_east, m_north);
    }

    private double m_north;
    private double m_south;
    private double m_east;
    private double m_west;
}